### ISSUE TITLE
Training from scratch on custom dataset

### ISSUE THREAD

**Issue by FenryrMKIII** [facebookresearch/detectron2#914]
I am reading the [paper ](https://arxiv.org/pdf/1811.08883.pdf) that has been linked in #348 in an attempt to train retinanet from scratch on my own dataset. I believe it's necessary in my case to train from scratch because I am attempting object detection on images that are quite different from imagenet (black & white, very few background). 

Based on my current understanding of the paper and my walk-through of detectron2 code I have performed the following modifications : 

```
cfg.INPUT.MIN_SIZE_TRAIN = (704, 1024, 1280)
cfg.INPUT.MAX_SIZE_TRAIN = (2048)

cfg.MODEL.WEIGHTS = ""
cfg.MODEL.PIXEL_MEAN = [218.96615195, 205.58776696, 199.45428186]
cfg.MODEL.PIXEL_STD = [20.1397195,  19.81115748, 21.38672478]
cfg.MODEL.ANCHOR_GENERATOR.SIZES = [[x, x * 2**(1.0/3), x * 2**(2.0/3) ] for x in [4, 8, 12, 16, 24 ]]
```
However when executing training, I also spot the following line in the network architecture : 

`(norm): FrozenBatchNorm2d(num_features=64, eps=1e-05)`

This is something that is directly discussed in the paper : 

> fine-tuning can adopt the pretraining batch statistics as fixed parameters [17]; however,
> freezing BN is invalid when training from scratch.

so for training from scratch, one should deactivate frozen batch normalization and move to either Group Normalization (GN) or Synchronized Batch Normalization (SyncBN) if using multiple GPUs (Note : that last bit is unclear for me : when I use google colab, can I use syncBN ? I assume not since I only have one GPU).

The question is : how to deactivate frozen batch normalization and activate GN in detectron2 ? 
Browsing through the code, I see there is a parameter `MODEL.BACKBONE.FREEZE_AT = 2` but it is unclear if this is what I should change ? What does this parameter do ? What does the 2 represent here in the resnet architecture ?

**Comment by FenryrMKIII** [comment]
Reading [this ](https://github.com/facebookresearch/detectron2/blob/master/MODEL_ZOO.md)page, I found a link towards [this ](https://github.com/facebookresearch/detectron2/blob/master/configs/Misc/scratch_mask_rcnn_R_50_FPN_3x_gn.yaml)model that is provided to train from scratch but which is a RCNN model. Assuming this can apply to the retinanet model, I made the following additions to my configuration : 

```
cfg.MODEL.BACKBONE.FREEZE_AT = 0
cfg.MODEL.RESNETS.NORM = "GN"
cfg.MODEL.FPN.NORM = "GN"
cfg.MODEL.ROI_BOX_HEAD = "GN"
```

**Comment by ppwwyyxx** [comment]
> when I use google colab, can I use syncBN ? 

you cannot.

> how to deactivate frozen batch normalization and activate GN in detectron2 

you can use `NORM="GN"` in various places in the config.
`ROI_BOX_HEAD="GN"` is irrelevant to RetinaNet.
The layers in RetinaNet currently do not have a "NORM" option and it contains no norm. We may add this in the future.

>  What does this parameter do ? What does the 2 represent here in the resnet architecture ?

"freeze" means not trained. So it should be 0.
2 means two stages - including a convolution and the first group of residual blocks.

**Comment by FenryrMKIII** [comment]
> > when I use google colab, can I use syncBN ?
> 
> you cannot.
> 
> > how to deactivate frozen batch normalization and activate GN in detectron2
> 
> you can use `NORM="GN"` in various places in the config.
> `ROI_BOX_HEAD="GN"` is irrelevant to RetinaNet.
> The layers in RetinaNet currently do not have a "NORM" option and it contains no norm. We may add this in the future.
> 
> > What does this parameter do ? What does the 2 represent here in the resnet architecture ?
> 
> "freeze" means not trained. So it should be 0.
> 2 means two stages - including a convolution and the first group of residual blocks.

Thanlks for your reply. When you say : 

> The layers in RetinaNet currently do not have a "NORM" option and it contains no norm. We may add this in the future.

What do you mean ? Because with the changes listed above, I see that the norm has changed in the network from FrozenBatchNorm to GroupNorm.  

With these changes, I now have a CUDA memory error altough I set very few images per batch : 

`cfg.SOLVER.IMS_PER_BATCH = 2`

My pictures are 704x1024 in the dataset. I allow data augmentation with following parameters : 

```
cfg.INPUT.MIN_SIZE_TRAIN = (704, 1024, 1280)
cfg.INPUT.MAX_SIZE_TRAIN = (2048)
```

So I see the following possibilities : 
- Use a smaller model 
- Reduce image size during augmentation
- Increase GPU

**Comment by ppwwyyxx** [comment]
> What do you mean ? Because with the changes listed above, I see that the norm has changed in the network from FrozenBatchNorm to GroupNorm.

To be more precise, I mean the layers in "RetinaNet Head" will not have Norm because the options are only about ResNet and FPN. Whether this will affect training is a different issue, though.

> Is this because when the model was frozen (cfg.MODEL.BACKBONE.FREEZE_AT = 2), memory consumption was somehow reduced 

Yes.
